encoding/custom/sema_codec/decode.go

```diff
@@ -0,0 +1,1069 @@
+// given io.Reader.
+func NewSemaDecoder(memoryGauge common.MemoryGauge, r io.Reader) *SemaDecoder {
+	return &SemaDecoder{
+		r:           LocatedReader{r: r},
+		typeDefs:    map[int]sema.Type{},
+		memoryGauge: memoryGauge,
+	}
+}
+
+// Decode reads custom-encoded bytes from the io.Reader and decodes them to a
+// Sema type. There is no assumption about the top-level Sema type so the first
+// byte must specify the top-level type. Usually this will be a CompositeType.
+//
+// This function returns an error if the bytes represent a custom encoding that
+// is malformed, does not conform to the custom specification, or contains
+// an unknown composite type.
+func (d *SemaDecoder) Decode() (t sema.Type, err error) {
+	return d.DecodeType()
+}
+
+func (d *SemaDecoder) DecodeElaboration() (el *sema.Elaboration, err error) {
+	el = sema.NewElaboration(d.memoryGauge, false)
+
+	err = DecodeMap(d, el.CompositeTypes, d.DecodeCompositeType)
+	if err != nil {
+		return
+	}
+
+	err = DecodeMap(d, el.InterfaceTypes, d.DecodeInterfaceType)
+	return
+}
+
+func (d *SemaDecoder) DecodeType() (t sema.Type, err error) {
+	typeIdentifier, err := d.DecodeTypeIdentifier()
+	if err != nil {
+		return
+	}
+
+	if isSimpleType(typeIdentifier) {
+		t, err = EncodingToSimpleType(typeIdentifier)
+	} else if isNumericType(typeIdentifier) {
+		t, err = EncodingToNumericType(typeIdentifier)
+	} else if isFixedPointNumericType(typeIdentifier) {
+		t, err = EncodingToFixedPointNumericType(typeIdentifier)
+	} else {
+		switch typeIdentifier {
+		case EncodedSemaOptionalType:
+			t, err = d.DecodeOptionalType()
+		case EncodedSemaReferenceType:
+			t, err = d.DecodeReferenceType()
+		case EncodedSemaCapabilityType:
+			t, err = d.DecodeCapabilityType()
+		case EncodedSemaVariableSizedType:
+			t, err = d.DecodeVariableSizedType()
+		case EncodedSemaAddressType:
+			t = &sema.AddressType{}
+		case EncodedSemaNilType:
+			t = nil
+
+		case EncodedSemaCompositeType:
+			t, err = d.DecodeCompositeType()
+		case EncodedSemaInterfaceType:
+			t, err = d.DecodeInterfaceType()
+		case EncodedSemaGenericType:
+			t, err = d.DecodeGenericType()
+		case EncodedSemaFunctionType:
+			t, err = d.DecodeFunctionType()
+		case EncodedSemaDictionaryType:
+			t, err = d.DecodeDictionaryType()
+		case EncodedSemaTransactionType:
+			t, err = d.DecodeTransactionType()
+		case EncodedSemaRestrictedType:
+			t, err = d.DecodeRestrictedType()
+		case EncodedSemaConstantSizedType:
+			t, err = d.DecodeConstantSizedType()
+
+		case EncodedSemaPointerType:
+			t, err = d.EncodePointer()
+
+		default:
+			err = fmt.Errorf("unknown type identifier: %d", typeIdentifier)
+		}
+	}
+
+	return
+}
+
+func (d *SemaDecoder) EncodePointer() (t sema.Type, err error) {
+	bufferOffset, err := d.DecodeLength()
+	if err != nil {
+		return
+	}
+
+	if knownType, defined := d.typeDefs[bufferOffset]; defined {
+		t = knownType
+	} else {
+		err = fmt.Errorf(`pointer to unknown type: %d`, bufferOffset)
```

Comment:
Could you please add tests for this and other uncovered code reported by codecov, especially the error/unhappy cases?